cm/config.py

```diff
@@ -16,186 +16,244 @@
+            return dict.__getitem__(self, "CM_" + key.upper())
+        else:
+            return dict.__getitem__(self, key)
+
+    def validate(self):
         # Check that required directories exist
-        for path in self.root, self.template_path:
+        for path in self.root_dir, self.template_path:
             if not os.path.isdir(path):
                 raise ConfigurationError(
                     "Directory does not exist: %s" % path)
 
-    def init_with_user_data(self, user_data):
-        self.__configure_instance_management(user_data)
-        self.__configure_instance_types(user_data)
-        self.condor_enabled = user_data.get("condor_enabled", False)
-        self.hadoop_enabled = user_data.get("hadoop_enabled", False)
-
-    def __configure_instance_management(self, user_data):
-        """Configure attributes used control reboot/terminate behavior
-        of cm.util.master:Instance."""
-        self.instance_reboot_timeout = \
-            user_data.get("instance_reboot_timeout", DEFAULT_INSTANCE_REBOOT_TIMEOUT)
-        self.instance_comm_timeout = \
-            user_data.get("instance_comm_timeout", DEFAULT_INSTANCE_COMM_TIMEOUT)
-        self.instance_state_change_wait = \
-            user_data.get("instance_state_change_wait", DEFAULT_INSTANCE_STATE_CHANGE_WAIT)
-        self.instance_reboot_attempts = \
-            user_data.get("instance_reboot_attempts", DEFAULT_INSTANCE_REBOOT_ATTEMPTS)
-        self.instance_terminate_attempts = \
-            user_data.get("instance_terminate_attempts", DEFAULT_INSTANCE_TERMINATE_ATTEMPTS)
-
-    def __configure_instance_types(self, user_data):
-        cloud_name = user_data.get('cloud_name', 'amazon').lower()
-        if "instance_types" in user_data:
+    @property
+    def user_data(self):
+        return self._user_data
+
+    @user_data.setter
+    def user_data(self, value):
+        self._user_data = value
+        self._rebuild_combined_config()
+
+    @property
+    def filesystem_templates(self):
+        if not 'filesystem_templates' in self:
+            self['filesystem_templates'] = self.app.manager.load_legacy_snapshot_data()
+        return self.get('filesystem_templates')
+
+    @property
+    def root_dir(self):
+        return self.get('root_dir', '.')
+
+    @property
+    def template_path(self):
+        """If 'path' is relative make absolute by prepending 'root'"""
+        template_path = self.get("template_path", "templates")
+        if not(os.path.isabs(template_path)):
+            path = os.path.join(self.root_dir, template_path)
+        return path
+
+    @property
+    def cloudman_source_file_name(self):
+        return self.get("cloudman_file_name", "cm.tar.gz")
+
+    @property
+    def cloud_name(self):
+        return self.get('cloud_name', 'amazon').lower()
+
+    @property
+    def cloud_type(self):
+        return self.get('cloud_type', 'ec2').lower()
+
+    @property
+    def multiple_processes(self):
```

Comment:
It would probably be better to keep this application specific than general, ie, renaming it so `galaxy_mutiple_processes` or similar.